File: mindspore/python/mindspore/_extends/graph_kernel/expanders/layernorm.py
# Copyright 2020-2022 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ===========================================================================
"""generate json desc for LayerNorm"""
from mindspore._extends.graph_kernel.model.model import DataFormat as DF
from ._utils import Expander, ExpanderInfoValidator as VLD
from ._utils import infer_shape_from_fractalnz, get_reduced_ori_shape, to_frac_z_axis


@VLD.add_format(DF.FRAC_NZ, DF.DEFAULT, DF.DEFAULT)
@VLD.add_format(DF.DEFAULT, DF.DEFAULT, DF.DEFAULT)
@VLD.check_attrs('begin_norm_axis', 'begin_params_axis', 'epsilon')
class LayerNorm(Expander):
    """LayerNorm expander"""

    def _expand(self, graph_builder):
        input_x, input_gamma, input_beta = self.inputs
        processor = self.processor
        begin_norm_axis = self.attrs['begin_norm_axis']
        epsilon = self.attrs['epsilon']

        ori_dtype = input_x.dtype
        if processor == 'aicore' and ori_dtype == 'float16':
            input_x = graph_builder.emit('Cast', [input_x], attrs={'dst_type': 'float32'})
            input_gamma = graph_builder.emit('Cast', [input_gamma], attrs={'dst_type': 'float32'})
            input_beta = graph_builder.emit('Cast', [input_beta], attrs={'dst_type': 'float32'})

        ori_shape_x = input_x.shape
        if input_x.data_format == DF.FRAC_NZ:
            ori_shape_x = infer_shape_from_fractalnz(input_x.shape)

        # Calculate the scaling ratio of the average
        if begin_norm_axis < 0:
            begin_norm_axis += len(ori_shape_x)

        reduce_axis = ()
        for i, _ in enumerate(ori_shape_x):
            if i > begin_norm_axis or i == begin_norm_axis:
                reduce_axis = reduce_axis + (i,)

        reduce_elts = 1.0
        for i in reduce_axis:
            reduce_elts *= ori_shape_x[i]
        # after reduced
        ori_reduced_shape_x = get_reduced_ori_shape(ori_shape_x, reduce_axis)

        axis = reduce_axis
        if input_x.data_format == DF.FRAC_NZ:
            axis = to_frac_z_axis(ori_shape_x, reduce_axis)

        mean_cof_v = graph_builder.value(input_x.dtype, 1.0 / reduce_elts)

        # Calculate mean
        mean_red = graph_builder.emit('ReduceSum', [input_x], attrs={'reduce_axis': axis, 'keep_dims': True})
        mean = graph_builder.emit('Mul', [mean_red, mean_cof_v])
        if input_x.data_format == DF.FRAC_NZ:
            mean = graph_builder.emit('Reshape', [mean], attrs={'shape': ori_reduced_shape_x})

        # Calculate variance
        variance_sub = graph_builder.emit('Sub', [input_x, mean])
        variance_mul = graph_builder.emit('Mul', [variance_sub, variance_sub])
        variance_red = graph_builder.emit('ReduceSum', [variance_mul], attrs={'reduce_axis': axis, 'keep_dims': True})
        variance = graph_builder.emit('Mul', [variance_red, mean_cof_v])
        if input_x.data_format == DF.FRAC_NZ:
            variance = graph_builder.emit('Reshape', [variance], attrs={'shape': ori_reduced_shape_x})

        # Calculate normalize
        normalize_sub = graph_builder.emit('Sub', [input_x, mean])
        epsilon_v = graph_builder.value(input_x.dtype, epsilon)
        normalize_add = graph_builder.emit('Add', [variance, epsilon_v])
        normlize_rsqrt = graph_builder.emit('Rsqrt', [normalize_add])
        normalize_mul = graph_builder.emit('Mul', [normalize_sub, normlize_rsqrt])

        # Calculate scale and translate
        scale_mul = graph_builder.emit('Mul', [normalize_mul, input_gamma])
        res = graph_builder.emit('Add', [scale_mul, input_beta])

        if processor == 'aicore' and ori_dtype == 'float16':
            res = graph_builder.emit('Cast', [res], attrs={'dst_type': 'float16'})
            mean = graph_builder.emit('Cast', [mean], attrs={'dst_type': 'float16'})
            variance = graph_builder.emit('Cast', [variance], attrs={'dst_type': 'float16'})
        return res, mean, variance

File: mindspore/ccsrc/backend/common/graph_kernel/expanders/layernorm.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <memory>
#include <vector>
#include "backend/common/graph_kernel/expanders/op_desc_registry.h"

namespace mindspore::graphkernel::expanders {
class LayerNorm : public OpDesc {
 public:
  LayerNorm() {
    auto support_format = std::make_unique<SupportFormat>();
    support_format->AddFormat({kOpFormat_FRAC_NZ, kOpFormat_DEFAULT, kOpFormat_DEFAULT});
    support_format->AddFormat({kOpFormat_DEFAULT, kOpFormat_DEFAULT, kOpFormat_DEFAULT});
    (void)validators_.emplace_back(std::move(support_format));
    std::initializer_list<std::string> attrs{"begin_norm_axis", "begin_params_axis", "epsilon"};
    (void)validators_.emplace_back(std::make_unique<CheckAttr>(attrs));
  }
  ~LayerNorm() = default;

 protected:
  NodePtrList Expand(const NodePtrList &inputs) override {
    const auto &input_x = inputs[0];
    const auto &input_gamma = inputs[1];
    const auto &input_beta = inputs[2];
    const auto epsilon = GetValue<float>(attrs_["epsilon"]);

    auto x = input_x;
    auto gamma = input_gamma;
    auto beta = input_beta;
    auto ori_type = input_x->type;
    if (processor_ == "aicore" && ori_type == TypeId::kNumberTypeFloat16) {
      x = gb.Cast(x, TypeId::kNumberTypeFloat32);
      gamma = gb.Cast(gamma, TypeId::kNumberTypeFloat32);
      beta = gb.Cast(beta, TypeId::kNumberTypeFloat32);
    }

    auto ori_shape_x = input_x->shape;
    if (input_x->format == kOpFormat_FRAC_NZ) {
      ori_shape_x = InferShapeFromFractalnz(ori_shape_x);
    }

    auto begin_norm_axis = GetValue<int64_t>(attrs_["begin_norm_axis"]);
    if (begin_norm_axis < 0) {
      begin_norm_axis += static_cast<int64_t>(ori_shape_x.size());
    }

    ShapeVector reduce_axis;
    int64_t reduce_elts = 1;
    for (size_t i = 0; i < ori_shape_x.size(); i++) {
      auto axis_int64 = SizeToLong(i);
      if (axis_int64 >= begin_norm_axis) {
        reduce_axis.push_back(axis_int64);
        reduce_elts *= ori_shape_x[i];
      }
    }

    auto ori_reduced_shape_x = GetReducedOriShape(ori_shape_x, reduce_axis);
    auto axis = reduce_axis;
    if (x->format == kOpFormat_FRAC_NZ) {
      axis = ToFracZAxis(ori_shape_x, reduce_axis);
    }

    auto mean_cof_tensor = gb.Tensor(1.0 / reduce_elts, x->type);

    // Calculate mean
    auto mean_red = gb.ReduceSum(x, axis, true);
    auto mean = gb.Mul(mean_red, mean_cof_tensor);
    if (x->format == kOpFormat_FRAC_NZ) {
      mean = gb.Reshape(mean, ori_reduced_shape_x);
    }

    // Calculate variance
    auto variance_sub = gb.Sub(x, mean);
    auto variance_mul = gb.Mul(variance_sub, variance_sub);
    auto variance_red = gb.ReduceSum(variance_mul, axis, true);
    auto variance = gb.Mul(variance_red, mean_cof_tensor);
    if (x->format == kOpFormat_FRAC_NZ) {
      variance = gb.Reshape(variance, ori_reduced_shape_x);
    }

    // Calculate normalize
    auto epsilon_tensor = gb.Tensor(epsilon, x->type);
    auto normalize_add = gb.Add(variance, epsilon_tensor);
    auto normalize_rsqrt = gb.Emit("Rsqrt", {normalize_add});
    auto normalize_mul = gb.Mul(variance_sub, normalize_rsqrt);

    // Calculate scale and translate
    auto scale_mul = gb.Mul(normalize_mul, gamma);
    auto res = gb.Add(scale_mul, beta);

    if (processor_ == "aicore" && ori_type == TypeId::kNumberTypeFloat16) {
      res = gb.Cast(res, TypeId::kNumberTypeFloat16);
      mean = gb.Cast(mean, TypeId::kNumberTypeFloat16);
      variance = gb.Cast(variance, TypeId::kNumberTypeFloat16);
    }
    return {res, mean, variance};
  }
};
EXPANDER_OP_DESC_REGISTER("LayerNorm", LayerNorm);
}  // namespace mindspore::graphkernel::expanders
